aiohttp/web_urldispatcher.py

```diff
@@ -998,27 +1008,75 @@ def __init__(self) -> None:
         self._named_resources: Dict[str, AbstractResource] = {}
         self._resource_index: dict[str, list[AbstractResource]] = {}
         self._matched_sub_app_resources: List[MatchedSubAppResource] = []
+        self._hyperdb: Optional[hyperscan.Database] = None  # type: ignore[no-any-unimported]
+        self._plain_resources: dict[str, PlainResource] = {}
+        self._prefix_resources: list[tuple[str, PrefixResource]] = []
+
+    def _on_match(
+        self, id_: int, from_: int, to: int, flags: int, found: list[int]
+    ) -> Optional[bool]:
+        found.append(id_)
+        return None
 
     async def resolve(self, request: Request) -> UrlMappingMatchInfo:
-        resource_index = self._resource_index
-        allowed_methods: Set[str] = set()
-
-        # Walk the url parts looking for candidates. We walk the url backwards
-        # to ensure the most explicit match is found first. If there are multiple
-        # candidates for a given url part because there are multiple resources
-        # registered for the same canonical path, we resolve them in a linear
-        # fashion to ensure registration order is respected.
-        url_part = request.rel_url.path_safe
-        while url_part:
-            for candidate in resource_index.get(url_part, ()):
-                match_dict, allowed = await candidate.resolve(request)
-                if match_dict is not None:
-                    return match_dict
-                else:
-                    allowed_methods |= allowed
-            if url_part == "/":
-                break
-            url_part = url_part.rpartition("/")[0] or "/"
+        allowed_methods: set[str] = set()
+        path = request.rel_url.path_safe
+
+        if (plain_resource := self._plain_resources.get(path)) is not None:
+            match_dict, allowed = await plain_resource.resolve(request)
+            if match_dict is not None:
+                return match_dict
+            else:
+                allowed_methods |= allowed
+
+        for prefix, prefix_resource in self._prefix_resources:
```

Comment:
I think this part can probably reuse the previous code with the `while url_part`, which would stop this growing in time as more prefixed resources are added.

Reply:
I'm experimented with alternative implementation for prefixed resources that should be faster than index dict, see updated PR.

Reply:
That looks like it's doubled the improvement on gitapi_subapps, so looking good to me.